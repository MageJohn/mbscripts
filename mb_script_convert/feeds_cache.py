import atexit
import pickle
from os import environ
from pathlib import Path
from typing import Any


class Cache[_KT, _VT](dict[_KT, _VT]):
    def __init__(self, *args, **kwargs):
        self.modified: bool = False
        return super().__init__(*args, **kwargs)

    def __setitem__(self, key: _KT, value: _VT, /) -> None:
        self.modified = True
        return super().__setitem__(key, value)


XDG_CACHE_HOME = Path(environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
CACHE_FILE = XDG_CACHE_HOME / "mb-scripts" / "cache.pickle"


if CACHE_FILE.exists():
    with CACHE_FILE.open("rb") as f:
        cache: Cache[str, Any] = pickle.load(f)
        cache.modified = False
else:
    cache = Cache()


def save_cache():
    if cache.modified:
        CACHE_FILE.parent.mkdir(parents=True, exist_ok=True)
        with CACHE_FILE.open("wb") as f:
            pickle.dump(cache, f, pickle.HIGHEST_PROTOCOL)


atexit.register(save_cache)
