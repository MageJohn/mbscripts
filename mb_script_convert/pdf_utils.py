from __future__ import annotations

import re
from math import isclose
from typing import TYPE_CHECKING

from py_pdf_parser.common import BoundingBox

if TYPE_CHECKING:
    from collections.abc import Callable

    from py_pdf_parser.filtering import PDFElement


def is_centered(element: PDFElement) -> bool:
    page = element.document.get_page(element.page_number)
    page_middle = page.width / 2

    bbox = element.bounding_box
    el_middle = (bbox.x0 + bbox.x1) / 2

    return isclose(page_middle, el_middle, rel_tol=0.015)


def is_in_top_right(element: PDFElement) -> bool:
    """
    If it is the top right of the page, could be a page number
    """
    bb = element.bounding_box
    return bb.x0 >= 500 and bb.y0 >= 740


def is_off_page(element: PDFElement) -> bool:
    page = element.document.get_page(element.page_number)
    page_bb = BoundingBox(0, page.width, 0, page.height)
    return not element.partially_within(page_bb)


def by_indent(indent: float | int) -> Callable[[PDFElement], bool]:
    def filter(element: PDFElement) -> bool:
        return isclose(element.bounding_box.x0, indent, abs_tol=10)

    return filter


def by_min_indent(indent: float | int) -> Callable[[PDFElement], bool]:
    def filter(element: PDFElement) -> bool:
        return element.bounding_box.x0 > indent

    return filter


def clean_text(element: PDFElement) -> str:
    return join_lines(element.text())


LINE_BREAK = re.compile(r" *\n *")


def join_lines(string: str) -> str:
    return LINE_BREAK.sub(" ", string)


class NotTaggedError(Exception):
    pass
