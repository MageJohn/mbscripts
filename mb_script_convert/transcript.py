from typing import Literal

type Tag = Literal["direction", "character", "parenthetical", "dialogue", "end"]
type Content = list[tuple[Tag, str]]
type Metadata = dict[str, str | Metadata]


class Transcript:
    """
    Represents a transcript in a format agnostic way.
    Used as an intermediate step when converting.
    """

    content: Content
    metadata: Metadata

    def __init__(self, metadata: Metadata = {}, content: Content = []):
        self.metadata = metadata
        self.content = content

    def add_content(self, tag: Tag, content: str):
        self.content.append((tag, content))

    def set_metadata(self, key: str, value: str):
        self.metadata[key] = value
