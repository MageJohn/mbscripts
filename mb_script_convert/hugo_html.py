from contextlib import nullcontext
from io import BytesIO
from pathlib import Path
from typing import IO

import lxml.html as html
import lxml.html.builder as b
import tomli_w

from .transcript import Metadata, Transcript

type HugoFrontmatter = dict[str, str | HugoFrontmatter]


def dump(transcript: Transcript, file_or_path: str | Path | IO[bytes]):
    if isinstance(file_or_path, str) or isinstance(file_or_path, Path):
        file = open(file_or_path, "wb+")
    else:
        file = nullcontext(file_or_path)
    with file as fh:
        fh.write(b"+++\n")
        tomli_w.dump(_metadata_to_hugo_frontmatter(transcript.metadata), fh)
        fh.write(b"+++\n\n")

        for el in transcript.content:
            html_el = b.P(b.CLASS(el[0]), el[1])
            html_el_str = html.tostring(html_el, pretty_print=True, encoding="utf-8")
            assert isinstance(html_el_str, bytes)  # for type hinting
            fh.write(html_el_str)


def dumps(transcript: Transcript) -> str:
    out = BytesIO()
    dump(transcript, out)
    return out.getvalue().decode("utf-8")


def _metadata_to_hugo_frontmatter(metadata: Metadata):
    frontmatter: HugoFrontmatter = {"params": {}}
    for k in metadata:
        if k == "episode-title":
            frontmatter["title"] = metadata[k]
        else:
            frontmatter["params"][k] = metadata[k]  # pyright: ignore [reportIndexIssue]
    return frontmatter
