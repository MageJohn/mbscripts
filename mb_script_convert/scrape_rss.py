import re
from urllib.parse import urlparse
from warnings import warn

import feedparser

from .feeds_cache import cache
from .transcript import Transcript


def scrape_episode_metadata(transcript: Transcript, url_file_stream_or_string):
    ep_title = transcript.metadata.episode_title
    assert (
        ep_title is not None
    ), "Cannot scrape episode metadata when episode title is not set"

    feed = _get_feed(url_file_stream_or_string)
    if feed is None:
        return

    episode = _match_episode(ep_title, feed)
    if episode is None:
        warn(
            f"Could not find episode with title {ep_title} in the RSS feed. Additional metadata has not been scraped"
        )
        return

    m = transcript.metadata
    m.cover_url = episode.image.href
    m.date_published = episode.published
    m.season = episode.itunes_season
    m.season_episode_number = episode.itunes_episode


def _get_feed(url_file_stream_or_string):
    looks_like_url = isinstance(url_file_stream_or_string, str) and urlparse(
        url_file_stream_or_string
    )[0] in ("http", "https")
    if not looks_like_url:
        return feedparser.parse(url_file_stream_or_string)
    url: str = url_file_stream_or_string

    if url not in cache:
        cache[url] = feedparser.parse(url)
        match cache[url].status:
            case 200 | 302:
                # all good
                pass
            case 301:
                warn(
                    f"HTTP response 301: {url} has permanently moved to {cache[url].href}"
                )
            case _:
                warn(
                    f"Something went wrong fetching {url} (status {cache[url].status}). Cannot scrape metadata"
                )
                return
    else:
        cached_feed = cache[url]
        new_feed = feedparser.parse(
            url, etag=cached_feed.get("etag"), modified=cached_feed.get("modified")
        )
        match new_feed.status:
            case 200 | 302:
                cache[url] = new_feed
            case 304:
                # the feed has not been modified
                pass
            case 301:
                warn(
                    f"HTTP response 301: {url} has permanently moved to {new_feed.href}"
                )
                cache[url] = new_feed
            case _:
                warn(
                    f"Something went wrong fetching {url} (status {new_feed.status}). Using cached feed."
                )

    return cache[url]


def _match_episode(ep_title: str, feed):
    title_re = re.compile(re.escape(ep_title), re.I)
    entry = None
    for entry in feed.entries:
        if title_re.search(entry.title):
            break
    return entry
