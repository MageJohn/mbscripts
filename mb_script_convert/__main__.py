import argparse
import sys

from mb_script_convert.scrape_rss import scrape_episode_metadata

from . import normalisations
from .hugo_html import dump
from .import_midnight_burger import import_transcript

MB_RSS_URL = "https://feeds.megaphone.fm/midnightburger"

parser = argparse.ArgumentParser()
parser.add_argument("infile")
parser.add_argument("-o", "--outfile", default=sys.stdout.buffer)
parser.add_argument(
    "--episode-title", help="Override the episode title as parsed from the PDF"
)
parser.add_argument(
    "--skip-scraping",
    action="store_true",
    help="Skip scraping extra metadata from the RSS feed",
)
parser.add_argument(
    "--rss-url",
    default=MB_RSS_URL,
    help="Override the URL of the Midnight Burger RSS feed",
)
parser.add_argument("--debug", action="store_true")


def main(infile, outfile, episode_title, skip_scraping, rss_url, debug):
    doc = import_transcript(infile, debug)
    normalisations.run_all(doc)
    if episode_title:
        doc.metadata.episode_title = episode_title
    if not skip_scraping:
        scrape_episode_metadata(doc, rss_url)
    dump(doc, outfile)


if __name__ == "__main__":
    args = parser.parse_args()
    main(
        args.infile,
        args.outfile,
        args.episode_title,
        args.skip_scraping,
        args.rss_url,
        args.debug,
    )
