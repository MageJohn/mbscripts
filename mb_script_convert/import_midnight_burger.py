import re
from warnings import warn

from py_pdf_parser.exceptions import ElementOutOfRangeError, NoElementFoundError
from py_pdf_parser.loaders import PDFDocument, load_file
from titlecase import titlecase

from mb_script_convert.pdf_utils import (
    by_indent,
    by_min_indent,
    clean_text,
    is_centered,
    is_in_top_right,
    is_off_page,
)

from .transcript import Transcript
from .transcript_utils import combine_more, extract_parentheticals

DIRECTIONS_INDENT = 108.0
DIALOGUE_INDENT = 144.0


def import_transcript(pdf_file: str, debug: bool) -> Transcript:
    pdf = load_file(pdf_file)
    tag_pdf(pdf)
    if debug:
        _visualise(pdf)
    transcript = tagged_pdf_to_transcript(pdf)
    extract_parentheticals(transcript)
    combine_more(transcript)
    return transcript


def _visualise(pdf: PDFDocument):
    from py_pdf_parser.visualise import visualise

    visualise(pdf)


def tag_pdf(pdf: PDFDocument):
    # Ignore elements off of the page
    pdf.elements.filter(is_off_page).ignore_elements()

    # Ignore page numbers
    page_numbers = pdf.elements.filter_by_regex(r"\d+\.").filter(is_in_top_right)
    if len(page_numbers) == 0:
        warn("Warning: could not find page numbers.")
    page_numbers.ignore_elements()

    # Detect and ignore title page
    first_page = pdf.get_page(1).elements
    if all(is_centered(el) for el in first_page) and len(first_page) <= 5:
        first_page.ignore_elements()
        first_page = pdf.elements.filter_by_page(2)
    else:
        warn("Warning: could not find title page")

    # Tag front matter
    first_page[0].add_tag("series_title")
    first_page[1].add_tag("episode_title")

    # Tag ending
    last_page = pdf.pages[-1].elements
    try:
        end = (
            last_page.filter_by_regex(r".*\bend\b.*", re.I)
            .filter(by_min_indent(DIALOGUE_INDENT + 10))
            .last()
        )
        end.add_tag("end")
    except NoElementFoundError:
        warn("Warning: could not find THE END or equivalent")

    # Rest of the script
    script = pdf.elements - pdf.elements.filter_by_tags(
        "series_title", "episode_title", "end"
    )

    directions = script.filter(by_indent(DIRECTIONS_INDENT))
    dialogue = script.filter(by_indent(DIALOGUE_INDENT))

    directions.add_tag_to_elements("direction/character")
    dialogue.add_tag_to_elements("dialogue")

    others = script - script.filter_by_tags("direction/character", "dialogue")
    if len(others) > 0:
        print("Warning: found text that has not been categorised:")
        for el in others:
            print(f"{repr(el.text())} {el.bounding_box}")

    script = directions | dialogue

    for el in directions:
        try:
            next = script.move_forwards_from(el)
        except ElementOutOfRangeError:
            next = None
        if next is not None and "dialogue" in next.tags:
            el.tags = set(["character"])
        else:
            el.tags = set(["direction"])


def tagged_pdf_to_transcript(tagged: PDFDocument) -> Transcript:
    transcript = Transcript()
    for el in tagged.elements:
        if "series_title" in el.tags:
            transcript.metadata.series = titlecase(clean_text(el))
        elif "episode_title" in el.tags:
            transcript.metadata.episode_title = clean_text(el)
        elif "direction" in el.tags:
            transcript.add_content("direction", clean_text(el))
        elif "character" in el.tags:
            transcript.add_content("character", clean_text(el))
        elif "dialogue" in el.tags:
            transcript.add_content("dialogue", clean_text(el))
        elif "end" in el.tags:
            transcript.add_content("end", clean_text(el))
    return transcript
