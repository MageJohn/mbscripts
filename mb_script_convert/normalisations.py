"""
Functions which normalise the text content of a Transcript
"""

from mb_script_convert.transcript import Transcript


def run_all(transcript: Transcript):
    title_period(transcript)


def title_period(transcript: Transcript):
    title = transcript.metadata.episode_title
    if title and title[-1] == ".":
        transcript.metadata.episode_title = title[:-1]
