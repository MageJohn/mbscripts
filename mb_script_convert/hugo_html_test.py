from pathlib import Path

import pytest

from mb_script_convert.hugo_html import dump, dumps
from mb_script_convert.transcript import Metadata, Transcript


@pytest.fixture
def transcript():
    return Transcript(
        metadata=Metadata(
            episode_title="Panopticon", season=4, date_published="October 17, 2019"
        ),
        content=[
            ("direction", "TAPE CLICKS ON"),
            ("character", "PETER"),
            ("parenthetical", "(pleasantly)"),
            ("dialogue", "Is everything alright, Martin?"),
        ],
    )


def test_dumps(transcript, snapshot):
    assert dumps(transcript) == snapshot


def test_dump(transcript, snapshot, tmp_path: Path):
    out_path = tmp_path / "out_Path.html"
    out_str = (tmp_path / "out_str.html").as_posix()
    out_fh_path = tmp_path / "out_fh.html"

    dump(transcript, out_path)
    dump(transcript, out_str)
    with open(out_fh_path, "wb") as out_fh:
        dump(transcript, out_fh)

    with open(out_path, "r") as fh:
        assert fh.read() == snapshot
    with open(out_str, "r") as fh:
        assert fh.read() == snapshot
    with open(out_fh_path, "r") as fh:
        assert fh.read() == snapshot
