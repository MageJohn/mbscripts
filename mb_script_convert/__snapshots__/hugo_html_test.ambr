# serializer version: 1
# name: test_dump
  '''
  +++
  title = "Panopticon"
  
  [params]
  season = 4
  date_published = "October 17, 2019"
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
# name: test_dump.1
  '''
  +++
  title = "Panopticon"
  
  [params]
  season = 4
  date_published = "October 17, 2019"
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
# name: test_dump.2
  '''
  +++
  title = "Panopticon"
  
  [params]
  season = 4
  date_published = "October 17, 2019"
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
# name: test_dumps
  '''
  +++
  title = "Panopticon"
  
  [params]
  season = 4
  date_published = "October 17, 2019"
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
